### PR TITLE
Feature/tcp talk respond

### PR DESCRIPTION
**Pull request by rudxor02**
**Feature for `Person` talking on tcp protocol**

now people have `Mouth`, `Ear` for each.

- In `Ear` there's listening socket which can receive message from others, more precisely another `Person`'s `Mouth`
  - port for listening socket is allocated in range ["PORT_START", "PORT_START" + "PORT_RANGE"), which can be set from env
  - if there's no port for socket, it raises exception (TODO)
  - if there's nothing to do for `Person`, `Ear` just keep listening
  - when `Ear` hear somethig, `Person` do `respond` to the message
  - when `Ear` hear something while `Person` doing something, `Person` finish that and do `respond` to the next message
  - all these processes are executed in another thread, which means now `Person` can `act` in non-blocking, not waiting for response
  - only user `Person` (David) `wait` for response, because we have to wait until `Civilization.solve` done
- `Mouth` talks to another `Person`'s `Ear`
- Message format (big endian) is 
  - instruction_length (4 bytes, uint)
  - extra_length (4 bytes, uint)
  - sender_name (12 bytes, utf)
  - message_type (4bytes, MessageType enum, uint)
    - now there's only one message type, `MessageType.Default`
  - instruction (`instruction_length` bytes, utf)
  - extra (`extra_length` bytes, utf)

### CODE REVIEW

File: core/civilization/person/base.py [hunk 1]
```diff
@@ -56,7 +60,21 @@ def to_format(self, message: str) -> str:
         )
 
 
-class BasePerson(BaseModel, PersonMessageFormat):
+class MessageType(Enum):
```
Comment (adldotori) [inline core/civilization/person/base.py:63]: Detail so good

File: core/civilization/person/base.py [hunk 2]
```diff
@@ -56,7 +60,21 @@ def to_format(self, message: str) -> str:
...
+    Default = 1
+
+
+INSTRUCTION_LENGTH_BYTES = 4
+EXTRA_LENGTH_BYTES = 4
+MESSAGE_SENTER_BYTES = 12
```
Comment (adldotori) [inline core/civilization/person/base.py:69]:
```suggestion
MESSAGE_SENDER_BYTES = 12
```

File: core/civilization/person/ear/base.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from socket import socket
+
+
+class MessageType(Enum):
```
Comment (adldotori) [inline core/civilization/person/ear/base.py:6]: It's duplicated

File: core/civilization/person/ear/default.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            try:
+                message_sender, _, instruction, extra = self.parse_data(conn)
+            except Exception as e:
+                print(
+                    System.announcement(
+                        f"[{self.person.name}-Ear] : Parsing Message Failed\n{e}"
+                    )
+                )
+                continue
+            self.person.respond(
```
Comment (adldotori) [inline core/civilization/person/ear/default.py:109]: How about handing this method over to callback?

File: core/civilization/person/default.py [hunk 5]
```diff
@@ -46,6 +51,7 @@ def respond(self, sender: Person, request: str, params: TalkParams) -> str:
             result, is_finish = self.execute(plans[0], sender=sender)
 
             if is_finish:
+                self.mouth.talk(sender.ear.port, result, "")
```
Comment (adldotori) [inline core/civilization/person/default.py:54]: mouth talks to ear. not ear's port. agree?
```suggestion
                self.mouth.talk(sender.ear, result, "")
```

**Review by adldotori — COMMENTED** [review]
so cool PR. I'm proud of you.

File: core/civilization/person/default.py (hunk 5 above)
Comment (rudxor02) [reply core/civilization/person/default.py:54]: ah yes yes

File: core/civilization/person/ear/base.py (hunk 3 above)
Comment (rudxor02) [reply core/civilization/person/ear/base.py:6]: it's `socket.socket`

File: core/civilization/person/ear/default.py (hunk 4 above)
Comment (rudxor02) [reply core/civilization/person/ear/default.py:109]: you mean `def listen(self, callback)`?

Comment (adldotori) [reply core/civilization/person/ear/default.py:109]: yes exactly

Comment (rudxor02) [reply core/civilization/person/ear/default.py:109]: It doesn't matter but is there any other places that can call `listen`? For now message format is only for `talk` and `respond`, as it contains `instruction`, `extra`. I think using callback means `Person` can listen differently to the message.

Comment (adldotori) [reply core/civilization/person/ear/default.py:109]: it's ambiguous. let's skip now.

**Review by adldotori — COMMENTED** [review]
could you test [this prompt](https://discord.com/channels/1102259622481498172/1104125261580091403/1104125261580091403)?

**Comment by rudxor02** [comment]
> could you test [this prompt](https://discord.com/channels/1102259622481498172/1104125261580091403/1104125261580091403)?

It works

**Review by adldotori — APPROVED** [review]

**Pull request merged**